src/lib.rs

```diff
@@ -158,11 +174,35 @@ impl<T> Drop for Pointer<T> {
 }
 
 
-// Warning: this exposes deleted entries
-impl<'a, T> ops::Deref for ReadLock<'a, T> {
-    type Target = [T];
-    fn deref(&self) -> &Self::Target {
-        &self.guard.data
+/// The item of `ReadIter`.
+pub struct ReadItem<'a, T: 'a> {
+    value: &'a T,
+    index: usize,
+}
+
+impl<'a, T> ops::Deref for ReadItem<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.value
+    }
+}
+
+impl<'a, T> Iterator for ReadIter<'a, T> {
+    type Item = ReadItem<'a, T>;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let id = self.index;
+            if id == self.lock.guard.data.len() {
+                return None
+            }
+            self.index += 1;
+            if !self.skip_lost || self.lock.guard.meta[id] != 0 {
```

Comment:
Do you think the compiler can move the branch out of the loop or should we have `ReadIterAlive`?